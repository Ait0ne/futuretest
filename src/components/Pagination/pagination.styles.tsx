import styled from 'styled-components';

export const PaginationContainer = styled.div`
display: flex;
width:100%;
align-items: center;
justify-content: flex-end;
margin-top:5px;
>div {
display: flex;
flex-direction: row;
justify-content: flex-end;
width:60%;
}
`
interface PageProps {
    disabled?: boolean,
    active?: boolean 
}

export const Page = styled.span<PageProps>`
cursor: ${props => props.disabled? 'default': 'pointer'};
color: ${props =>  props.disabled? 'rgba(0,0,0,0.5)': 'black'};
font-size: 1.2rem;
font-weight: ${props=> props.active? 'bold': '400'};
margin-right: 5px;
`