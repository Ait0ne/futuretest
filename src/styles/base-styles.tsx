import styled from 'styled-components';

interface ButtonProps {
    active?: boolean
}

export const Button = styled.button<ButtonProps>`
cursor: pointer;
min-width: 64px;
padding: 6px 16px;
color: rgba(0, 0, 0, 0.87);
text-transform: uppercase;
font-weight: 500;
border-radius:4px;
border: none;
margin: 5px;
background-color: #e0e0e0;
box-shadow: 0px 3px 1px -2px rgba(0,0,0,0.2), 0px 2px 2px 0px rgba(0,0,0,0.14), 0px 1px 5px 0px rgba(0,0,0,0.12);
&:disabled {
    color: rgba(0, 0, 0, 0.26)
}
`

export const Input = styled.input`
padding: 12px 25px 12px 14px;
border-radius: 4px;
width:25ch;
border: 1px solid rgba(0,0,0,0.3);
font-size: 1rem;
margin: 5px 10px;
&:focus {
    outline:none;
}

`