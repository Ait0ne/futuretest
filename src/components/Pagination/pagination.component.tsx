import React, {useEffect, useState} from 'react';


import {PaginationContainer, Page} from './pagination.styles';

export interface PaginationProps {
    pagination: {
        perPage:number,
        currentPage: number
    },
    rowNumber: number,
    setCurrentPage: (page:number) => void
}

const Pagination: React.FC<PaginationProps> = ({pagination, rowNumber, setCurrentPage}) => {
    const numberOfPages = Math.ceil(rowNumber/pagination.perPage)
    
    const [displayedPages, setDisplayedPages] = useState<number[]>([])
    
    useEffect(()=> {
        if (numberOfPages<6) {
            setDisplayedPages(Array.from(Array(numberOfPages).keys()))
        } else if (pagination.currentPage<4) {
            setDisplayedPages([0,1,2,3,4])
        } else if (pagination.currentPage>numberOfPages-5) {
            const newPages = []
            for (let i=5; i>0; i--) {
                newPages.push(numberOfPages-i)
            }
            setDisplayedPages(newPages)
        } else {
            const {currentPage} = pagination
            setDisplayedPages([currentPage-2,currentPage-1,currentPage,currentPage+1, currentPage+2])
        }
    }, [pagination, rowNumber, numberOfPages])


    return(
        <PaginationContainer>
            <div>
                <Page onClick={()=> setCurrentPage(0)}>&#171;</Page>
                <Page 
                disabled={pagination.currentPage===0}
                onClick={pagination.currentPage===0?undefined : () => setCurrentPage(pagination.currentPage-1)}
                >
                    &#8249;
                </Page>
                {
                    displayedPages.map((page, index) => {
                        return (
                            <Page
                            key={index}
                            onClick={()=> setCurrentPage(page)}
                            active={pagination.currentPage===page}
                            >
                                {page+1}
                            </Page>
                        )
                    })
                }
                <Page 
                disabled={pagination.currentPage===numberOfPages-1}
                onClick={pagination.currentPage===numberOfPages-1?undefined : () => setCurrentPage(pagination.currentPage+1)}
                >
                    &#8250;
                </Page>
                <Page onClick={()=> setCurrentPage(numberOfPages-1)} >&#187;</Page>
            </div>
        </PaginationContainer>
    )
}

export default Pagination;