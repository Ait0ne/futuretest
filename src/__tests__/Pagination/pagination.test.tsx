import React from 'react';
import {shallow, mount} from 'enzyme';

import Pagination, {PaginationProps} from '../../components/Pagination/pagination.component';
import {Page} from '../../components/Pagination/pagination.styles';

const props:PaginationProps = {
    rowNumber: 500,
    setCurrentPage: jest.fn(),
    pagination: {
        currentPage:0,
        perPage:50
    }
} 



describe('Pagination', () => {
    const pagination = mount(<Pagination {...props}/>)
    const navigationButtons = pagination.find(Page)
    const pageNumbers = navigationButtons.slice(2,7)
    beforeAll(()=> {
        jest.spyOn(React, 'useEffect').mockImplementation(React.useLayoutEffect)
    })
    it('renders correctly', () => {
        expect(pagination).toMatchSnapshot()
    })

    it('renders 9 navigation buttons', () => {
        expect(navigationButtons.length).toEqual(9)
    })

    pageNumbers.forEach((page, index)=> {
        it('renders correct page number', () => {
            expect(page.text()).toEqual(`${index+1}`)
        })
    })
    
    

    it('sets active prop on current page', () => {
        const currentPage = pagination.find(Page).at(2)
        expect(currentPage.prop('active')).toEqual(true)
    })

    it('calls setCurrentPage function with correct values on page click', () => {
        const page = pagination.find(Page).at(4)
        page.simulate('click')
        
        expect(props.setCurrentPage).toHaveBeenCalledWith(2)
    })

    it('calls setCurrentPage function with correct values on go to first page button click', () => {
        const page = pagination.find(Page).at(0)
        page.simulate('click')

        expect(props.setCurrentPage).toHaveBeenCalledWith(0)
    })

    it('calls setCurrentPage function with correct values on go to last page button click', () => {
        const page = pagination.find(Page).at(8)
        page.simulate('click')

        expect(props.setCurrentPage).toHaveBeenCalledWith(props.rowNumber/props.pagination.perPage-1)
    })
})

describe('pagination with different props', () => {
    it('renders 6 navigation buttons for 100 rows of data', () => {
        const pagination = mount(<Pagination {...{...props, rowNumber:100}}/>)
        expect(pagination.find(Page).length).toEqual(6)
    })

    it('shows correct page numbers, when current page is set to 18', () => {
        const pagination = mount(<Pagination {...{...props, rowNumber:950, pagination:{currentPage:18, perPage: 50}}}/>)
        const pages = pagination.find(Page)
        expect(pages.at(5).text()).toEqual("18")
    })

    it('shows correct page numbers, when current page is set to 10', () => {
        const pagination = mount(<Pagination {...{...props, pagination:{currentPage:5, perPage: 50}}}/>)
        const pages = pagination.find(Page)
        expect(pages.at(4).text()).toEqual("6")
    })
})