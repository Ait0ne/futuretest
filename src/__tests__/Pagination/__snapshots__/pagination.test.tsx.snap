// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pagination renders correctly 1`] = `
<Pagination
  pagination={
    Object {
      "currentPage": 0,
      "perPage": 50,
    }
  }
  rowNumber={500}
  setCurrentPage={[MockFunction]}
>
  <styled.div>
    <div
      className="sc-bdnylx cRWxg"
    >
      <div>
        <styled.span
          onClick={[Function]}
        >
          <span
            className="sc-gtssRu kkeeNc"
            onClick={[Function]}
          >
            «
          </span>
        </styled.span>
        <styled.span
          disabled={true}
        >
          <span
            className="sc-gtssRu lltIXf"
            disabled={true}
          >
            ‹
          </span>
        </styled.span>
        <styled.span
          active={true}
          key="0"
          onClick={[Function]}
        >
          <span
            className="sc-gtssRu gGJAup"
            onClick={[Function]}
          >
            1
          </span>
        </styled.span>
        <styled.span
          active={false}
          key="1"
          onClick={[Function]}
        >
          <span
            className="sc-gtssRu kkeeNc"
            onClick={[Function]}
          >
            2
          </span>
        </styled.span>
        <styled.span
          active={false}
          key="2"
          onClick={[Function]}
        >
          <span
            className="sc-gtssRu kkeeNc"
            onClick={[Function]}
          >
            3
          </span>
        </styled.span>
        <styled.span
          active={false}
          key="3"
          onClick={[Function]}
        >
          <span
            className="sc-gtssRu kkeeNc"
            onClick={[Function]}
          >
            4
          </span>
        </styled.span>
        <styled.span
          active={false}
          key="4"
          onClick={[Function]}
        >
          <span
            className="sc-gtssRu kkeeNc"
            onClick={[Function]}
          >
            5
          </span>
        </styled.span>
        <styled.span
          disabled={false}
          onClick={[Function]}
        >
          <span
            className="sc-gtssRu kkeeNc"
            disabled={false}
            onClick={[Function]}
          >
            ›
          </span>
        </styled.span>
        <styled.span
          onClick={[Function]}
        >
          <span
            className="sc-gtssRu kkeeNc"
            onClick={[Function]}
          >
            »
          </span>
        </styled.span>
      </div>
    </div>
  </styled.div>
</Pagination>
`;
